src/NHSD.BuyingCatalogue.Solutions.Application/Queries/GetPricingBySolutionId/GetPriceBySolutionIdHandler.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using AutoMapper;
+using MediatR;
+using NHSD.BuyingCatalogue.Solutions.Application.Domain.Pricing;
+using NHSD.BuyingCatalogue.Solutions.Application.Persistence;
+using NHSD.BuyingCatalogue.Solutions.Contracts.Pricing;
+using NHSD.BuyingCatalogue.Solutions.Contracts.Queries;
+
+namespace NHSD.BuyingCatalogue.Solutions.Application.Queries.GetPricingBySolutionId
+{
+    internal sealed class GetPriceBySolutionIdHandler : IRequestHandler<GetPriceBySolutionIdQuery, IEnumerable<ICataloguePrice>>
+    {
+        private readonly PriceReader _priceReader;
+        private readonly IMapper _mapper;
+
+        public GetPriceBySolutionIdHandler(PriceReader priceReader, IMapper mapper)
+        {
+            _priceReader = priceReader;
+            _mapper = mapper;
+        }
+
+        public async Task<IEnumerable<ICataloguePrice>> Handle(GetPriceBySolutionIdQuery request, CancellationToken cancellationToken)
+        {
+            var prices = await _priceReader.GetBySolutionIdAsync(request.SolutionId, cancellationToken);
+
+            List<ICataloguePrice> cataloguePrices = new List<ICataloguePrice>();
+
+            foreach (var price in prices)
+            {
+                if (price is CataloguePriceFlat cataloguePriceFlat)
+                {
+                    cataloguePrices.Add(new FlatCataloguePriceDto
+                    {
+                        CataloguePriceId = price.CataloguePriceId,
+                        CatalogueItemName = price.CatalogueItemName,
+                        CatalogueItemId = price.CatalogueItemId,
+                        Type = price.CataloguePriceType.Name,
+                        CurrencyCode = price.CurrencyCode,
+                        PricingUnit = _mapper.Map<IPricingUnit>(price.PricingUnit),
+                        TimeUnit = _mapper.Map<ITimeUnit>(price.TimeUnit),
+                        Price = cataloguePriceFlat.Price
+                    });
+                }
+                else if (price is CataloguePriceTier cataloguePriceTier)
```

Comment:
I think we should try to avoid testing for a particular type and prefer polymorphism where possible. For example, an alternative here would be to add a member to `CataloguePriceBase` such as `IsFlat`. There may well be a better option – this was just a quick suggestion.

There's a quote from Scott Meyer's Effective C++ (although it applies to many languages) that I rather like:

> Anytime you find yourself writing code of the form 'if the object is of type T1, then do something, but if it's of type T2, then do something else,' slap yourself.

P.S. No offence intended!

Reply:
Made a tech debt ticket for this: https://buyingcatalog.visualstudio.com/Buying%20Catalogue/_sprints/taskboard/Buying%20Catalogue%20Tech%20Cell/Buying%20Catalogue/Iteration%2027%20Stretch%20Goals?workitem=8004